### DISCUSSION TITLE
kvBuckets are unexpectedly slow

### DISCUSSION THREAD

**Discussion by EgusCpp** [nats-io/nats-server#5334] [Q&A] [answered]
Greetings!  
I am using NATS kvBuckets with FileStorage. Mostly to store key-value pair and delete it over few seconds. It was all good with performance, until it reached over 100k pairs. Write and erase operations were still fine, while listing keys ("kv ls bucket_name") or clearing memory ("kv compact") started taking enormous time, provided that nats-server process consuming 98-99% of cpu core it attached to. Literally, before calling "kv compact" I had 106,538 records using 16MB of disk space, and after waiting 30+ minutes, I terminated nats-cli and now I have 85,904 records using 13MB of disk space. I.e. so much time to free 3MB? I don't understand what is wrong here, maybe it is not supposed to use buckets with file storage having big number of key-value pairs?  

Nats server v.2.10.7  

  
upd: 48+ minutes to "compact" the rest  
![image](https://github.com/nats-io/nats-server/assets/115817065/209318c5-63d8-4989-8488-2799e7b9249a)

**Comment by ripienaar** [comment]
Our design is heavily optimised for Get, Put, Update, Create, Delete etc being fast as possible. Listing keys is particularly slow in how it is currently designed as that is a client side operation.

We have some work planned to make that faster but for now listing keys is not fast and probably something that in typical use cases should not be done that often.

Likewise compact has to essentially scan through every key, do some checks etc, all on the client.  This is also a slow process inherently but again probably not something we'd imagine people do super commonly.

**Reply by EgusCpp** [reply]
Well, in our scenario I was trying to use kvBuckets as fast storage for failure recovery. I.e. store critical data, do something and erase data when ownership is transferred to another service. On service failure and after restart, we have to read what we have left in this bucket. And it was suprise: despite we had only like ~30 "real" key-value pairs, they were read after 6+ minutes of 100% CPU consumption! Probably because there were 100k+ records purged while not "freed" by "compact". It seems we have to call "compact" constantly to clean things before we meet enormous sizes...

**Reply by ripienaar** [reply]
Do you store your data in some hierarchy where you could delete keys based on a wildcard?

You say you have 30 keys but 100k pairs is this due to delete markers? Do you have a top end for how long you'd keep this data would a max age help?

**Reply by EgusCpp** [reply]
@ripienaar It is not supposed to delete keys by wildcard in our case, only per key individually. We have some kind of hierarchy: I just made separate buckets per each... let's say "type". And there is no time-based deletion.  
Yes, 30 keys is current "active" amount, and it could be up to 1-2k per "type", i.e. per "bucket". While "total" amount of keys is unlimited and may only grow. I.e. every new message means new key, and there shouldn't be same key never again after that.

**Reply by EgusCpp** [reply]
Maybe there is a way to reuse purged pairs for new unique keys, and I am just missing something?

**Reply by ripienaar** [reply]
Delete in KV is still slow as it has to write each key and mark it as deleted for audit trail purposes.

Well using a hierarchy will mean you can very quickly purge entire sub trees out so I would suggest considering that.  Our operations tend to be on nats wildcards from subjects so think about those.

Other option is to use a stream directly where you have more control of implementation specifics such as optimising deletes that don't write audit trails and purging out entire subsets of data in a way that means they wont need to call compact.

**Reply by EgusCpp** [reply]
And there is no way for kvBuckets to disable "audit trails" , history of changes, etc?

**Reply by ripienaar** [reply, marked as answer]
No, this is essentially what they are and it also serve another purpose that of making watch be aware of delete, we cant really turn that off.

Using a stream seems better for you

**Reply by EgusCpp** [reply]
Thank you for assistance! However, streams are also a bit different from behaviour expected from plain persistent key-value storage.

**Reply by ripienaar** [reply]
In what way?

**Reply by EgusCpp** [reply]
I would expect the following operations from persistent key-value storage: put key with value, get value for key, list available keys and delete key with value. There is no "consumer" entity supposed. Also NATS streams are providing messages deletion only by sequence number, as far as I can see, what is very incovenient when you need to address using only your own keys.

**Reply by ripienaar** [reply]
We have:

 * Put key with value
 * Get value for key
 * List available keys (albeit its a bit slow)
 * Delete key by key name not value (purge per subject which includes wildcards hence fast wildcard deletes of many keys)

I've not seen before a `delete key with value` in other KVs or do you mean only delete by key if the value is a certain value?

A KV bucket is a stream, so you can assume the stuff the KV API can do is also supported by streams.

**Reply by ripienaar** [reply]
For list keys, if you were to use a stream and do not care about history/audit, it could be essentially instantaneous via the stream info api.

**Reply by EgusCpp** [reply]
@ripienaar Yes, I was speaking about delete by key of course. Just wanted to say we need to delete both key and value during this operation leaving nothing behind.  
  
In other words, do you suggest to use subject as a key, so basically it is normal to have thouthands of subjects where each subject has only one message? And this method may work even when there is no consumers/subscribers?

**Reply by ripienaar** [reply]
Yes thousands of subjects are not a problem, this is how KV works.  KV Buckets are like `$KV.<BUCKET>.>` with keys below that. To delete you can do a purge operation like purge `$KV.BUCKET.prefix.>` and this will quite rapidly delete many thousands of messages below that prefix - this is why I suggested earlier you consider that for related messages.

In 2.11 we'll have optimised GET operations also that can get many subjects at a time in batches

Listing keys is a stream info operation with subject filter - this will then give you information for individual subjects up to 100k subjects - and it supports paging.  This is also super quick. But it's not aware of KV audit trail hence KV key list being slow as we have to check each key for deleted or not.

So with a little bit of use of the stream API you can tailor something really nice and fast.  KV is nice but it has a set use case in mind and when you find the KV abstraction is a bit not right you can go to normal streams and there you have many more capabilities.

**Reply by EgusCpp** [reply]
@ripienaar Sounds good, while what about necessity to subscribe? I cannot just start publishing subjects to the stream, as I am having error 34 (NATS_NO_RESPONDERS). I.e. I have to have a subscription, even if I am not interested in that?

**Reply by ripienaar** [reply]
You should not get no responders from a stream unless the stream is unhealthy so not sure where that is from. The whole purpose of the stream concept is to capture things even when subscribers are gone :)
